File: electron/sqlite/queries/partners.ts
/**
 * 合作/监管单位相关 SQLite 查询模块
 *
 * 实现 partners、regions、supervisors 三张表的 CRUD 操作。
 */

import log from 'electron-log'
import { tryGetSqlite, rowToCamel, toSqliteValue } from './helpers'

// ═══════════════════════════════════════════════════════════════════════════════
// Partners — 列映射
// ═══════════════════════════════════════════════════════════════════════════════

const PART_COLUMNS: Record<string, string> = {
  id: 'id',
  name: 'name',
  category: 'category',
  contact: 'contact',
  phone: 'phone',
  email: 'email',
  address: 'address',
  bankAccount: 'bank_account',
  bankName: 'bank_name',
  taxNumber: 'tax_number',
  creditCode: 'credit_code',
  registeredAddress: 'registered_address',
  businessScope: 'business_scope',
  taxType: 'tax_type',
  licenseFile: 'license_file',
  licenseFileType: 'license_file_type',
  otherFiles: 'other_files',
  otherFilesType: 'other_files_type',
  projectIds: 'project_ids',
  remarks: 'remarks',
  createdAt: 'created_at',
  updatedAt: 'updated_at',
}

const PART_INSERT_COLS = Object.values(PART_COLUMNS).filter(c => c !== 'id')
const PART_INSERT_SQL = `INSERT INTO partners (${PART_INSERT_COLS.map(c => `"${c}"`).join(', ')}) VALUES (${PART_INSERT_COLS.map(() => '?').join(', ')})`

// ═══════════════════════════════════════════════════════════════════════════════
// Regions — 列映射
// ═══════════════════════════════════════════════════════════════════════════════

const REG_COLUMNS: Record<string, string> = {
  id: 'id',
  province: 'province',
  city: 'city',
  district: 'district',
  createdAt: 'created_at',
}

const REG_INSERT_COLS = Object.values(REG_COLUMNS).filter(c => c !== 'id')
const REG_INSERT_SQL = `INSERT INTO regions (${REG_INSERT_COLS.map(c => `"${c}"`).join(', ')}) VALUES (${REG_INSERT_COLS.map(() => '?').join(', ')})`

// ═══════════════════════════════════════════════════════════════════════════════
// Supervisors — 列映射
// ═══════════════════════════════════════════════════════════════════════════════

const SUP_COLUMNS: Record<string, string> = {
  id: 'id',
  regionId: 'region_id',
  name: 'name',
  category: 'category',
  contact: 'contact',
  phone: 'phone',
  address: 'address',
  projectIds: 'project_ids',
  remarks: 'remarks',
  createdAt: 'created_at',
  updatedAt: 'updated_at',
}

const SUP_INSERT_COLS = Object.values(SUP_COLUMNS).filter(c => c !== 'id')
const SUP_INSERT_SQL = `INSERT INTO supervisors (${SUP_INSERT_COLS.map(c => `"${c}"`).join(', ')}) VALUES (${SUP_INSERT_COLS.map(() => '?').join(', ')})`

// ═══════════════════════════════════════════════════════════════════════════════
// 辅助
// ═══════════════════════════════════════════════════════════════════════════════

function toInsertParams(columns: Record<string, string>, insertCols: string[], obj: Record<string, any>): any[] {
  return insertCols.map(col => {
    const jsonKey = Object.entries(columns).find(([, c]) => c === col)?.[0]
    if (!jsonKey) return null
    return toSqliteValue(obj[jsonKey])
  })
}

function toUpdateSet(columns: Record<string, string>, changes: Record<string, any>, excludeKeys: string[] = []): { sql: string; params: any[] } {
  const setClauses: string[] = []
  const params: any[] = []
  for (const [jsonKey, value] of Object.entries(changes)) {
    if (excludeKeys.includes(jsonKey)) continue
    const col = columns[jsonKey]
    if (!col) continue
    setClauses.push(`"${col}" = ?`)
    params.push(toSqliteValue(value))
  }
  return { sql: setClauses.join(', '), params }
}

/** 从 project_ids JSON 数组解析项目名称 */
function resolveProjectNames(projectIdsJson: string | null, sqlite: any): string {
  if (!projectIdsJson) return ''
  try {
    const ids: number[] = JSON.parse(projectIdsJson)
    if (!ids.length) return ''
    const names = ids.map(id => {
      const row = sqlite.prepare('SELECT name FROM projects WHERE id = ?').get(id) as { name: string } | undefined
      return row?.name || ''
    }).filter(Boolean)
    return names.join(', ')
  } catch {
    return ''
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Partners — 操作
// ═══════════════════════════════════════════════════════════════════════════════

export function listPartners(): any[] | null {
  const sqlite = tryGetSqlite()
  if (!sqlite) return null
  try {
    const rows = sqlite.prepare('SELECT * FROM partners ORDER BY created_at DESC').all() as Record<string, any>[]
    return rows.map(row => {
      const camel = rowToCamel(row)
      camel.projectNames = resolveProjectNames(row.project_ids, sqlite)
      return camel
    })
  } catch (err) {
    log.error('[SQLite] partners.list error:', err)
    return null
  }
}

export function createPartner(partner: any): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const params = toInsertParams(PART_COLUMNS, PART_INSERT_COLS, partner)
    sqlite.prepare(PART_INSERT_SQL).run(...params)
    return true
  } catch (err) {
    log.error('[SQLite] partners.create error:', err)
    return false
  }
}

export function updatePartner(partner: any): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const { sql: setSql, params: setParams } = toUpdateSet(PART_COLUMNS, partner, ['id'])
    if (!setSql) return true
    setParams.push(new Date().toISOString())
    setParams.push(partner.id)
    const result = sqlite.prepare(`UPDATE partners SET ${setSql}, "updated_at" = ? WHERE id = ?`).run(...setParams)
    return result.changes > 0
  } catch (err) {
    log.error('[SQLite] partners.update error:', err)
    return false
  }
}

export function deletePartner(id: number): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    sqlite.prepare('DELETE FROM partners WHERE id = ?').run(id)
    return true
  } catch (err) {
    log.error('[SQLite] partners.delete error:', err)
    return false
  }
}

export function listPartnersByProject(projectId: number): any[] | null {
  const sqlite = tryGetSqlite()
  if (!sqlite) return null
  try {
    // project_ids 是 JSON 数组，需要用 LIKE 粗筛 + JS 精筛
    const rows = sqlite.prepare('SELECT * FROM partners').all() as Record<string, any>[]
    const filtered = rows.filter(row => {
      try {
        const ids: number[] = JSON.parse(row.project_ids || '[]')
        return ids.includes(projectId)
      } catch { return false }
    })
    return filtered.map(row => {
      const camel = rowToCamel(row)
      camel.projectNames = resolveProjectNames(row.project_ids, sqlite)
      return camel
    })
  } catch (err) {
    log.error('[SQLite] partners.listByProject error:', err)
    return null
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Regions — 操作
// ═══════════════════════════════════════════════════════════════════════════════

export function listRegions(): any[] | null {
  const sqlite = tryGetSqlite()
  if (!sqlite) return null
  try {
    const rows = sqlite.prepare('SELECT * FROM regions ORDER BY created_at DESC').all() as Record<string, any>[]
    return rows.map(rowToCamel)
  } catch (err) {
    log.error('[SQLite] regions.list error:', err)
    return null
  }
}

export function createRegion(region: any): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const params = toInsertParams(REG_COLUMNS, REG_INSERT_COLS, region)
    sqlite.prepare(REG_INSERT_SQL).run(...params)
    return true
  } catch (err) {
    log.error('[SQLite] regions.create error:', err)
    return false
  }
}

/** 检查地区是否被监管单位引用 */
export function countRegionRefs(regionId: number): number | null {
  const sqlite = tryGetSqlite()
  if (!sqlite) return null
  try {
    const row = sqlite.prepare('SELECT COUNT(*) as count FROM supervisors WHERE region_id = ?').get(regionId) as { count: number }
    return row.count
  } catch (err) {
    log.error('[SQLite] regions.countRefs error:', err)
    return null
  }
}

export function deleteRegion(id: number): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    sqlite.prepare('DELETE FROM regions WHERE id = ?').run(id)
    return true
  } catch (err) {
    log.error('[SQLite] regions.delete error:', err)
    return false
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Supervisors — 操作
// ═══════════════════════════════════════════════════════════════════════════════

export function listSupervisors(): any[] | null {
  const sqlite = tryGetSqlite()
  if (!sqlite) return null
  try {
    const rows = sqlite.prepare(`
      SELECT s.*,
        CASE WHEN r.id IS NOT NULL THEN r.province || '-' || r.city || '-' || r.district ELSE '' END as region_name
      FROM supervisors s
      LEFT JOIN regions r ON s.region_id = r.id
      ORDER BY s.created_at DESC
    `).all() as Record<string, any>[]
    return rows.map(row => {
      const camel = rowToCamel(row)
      camel.regionName = (row as any).region_name || ''
      camel.projectNames = resolveProjectNames(row.project_ids, sqlite)
      return camel
    })
  } catch (err) {
    log.error('[SQLite] supervisors.list error:', err)
    return null
  }
}

export function createSupervisor(supervisor: any): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const params = toInsertParams(SUP_COLUMNS, SUP_INSERT_COLS, supervisor)
    sqlite.prepare(SUP_INSERT_SQL).run(...params)
    return true
  } catch (err) {
    log.error('[SQLite] supervisors.create error:', err)
    return false
  }
}

export function updateSupervisor(supervisor: any): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const { sql: setSql, params: setParams } = toUpdateSet(SUP_COLUMNS, supervisor, ['id'])
    if (!setSql) return true
    setParams.push(new Date().toISOString())
    setParams.push(supervisor.id)
    const result = sqlite.prepare(`UPDATE supervisors SET ${setSql}, "updated_at" = ? WHERE id = ?`).run(...setParams)
    return result.changes > 0
  } catch (err) {
    log.error('[SQLite] supervisors.update error:', err)
    return false
  }
}

export function deleteSupervisor(id: number): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    sqlite.prepare('DELETE FROM supervisors WHERE id = ?').run(id)
    return true
  } catch (err) {
    log.error('[SQLite] supervisors.delete error:', err)
    return false
  }
}

File: src-tauri/src/commands/partners.rs
/**
 * 合作伙伴管理命令
 *
 * 对应 Electron 版本的 partners.ts
 */

use crate::db::AppState;
use crate::error::{AppError, AppResult};
use rusqlite::params;
use serde::{Deserialize, Serialize};
use tauri::{command, State};

/// 解析 JSON 字符串为 Vec<i64>
fn parse_ids(json: Option<String>) -> Option<Vec<i64>> {
    json.and_then(|s| serde_json::from_str(&s).ok())
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Partner {
    pub id: i64,
    pub name: String,
    pub category: Option<String>,
    pub contact: Option<String>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub address: Option<String>,
    pub bank_account: Option<String>,
    pub bank_name: Option<String>,
    pub tax_number: Option<String>,
    pub credit_code: Option<String>,
    pub registered_address: Option<String>,
    pub business_scope: Option<String>,
    pub tax_type: Option<String>,
    pub license_file: Option<String>,
    pub license_file_type: Option<String>,
    pub other_files: Option<String>,
    pub other_files_type: Option<String>,
    pub project_ids: Option<Vec<i64>>,
    pub remarks: Option<String>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreatePartner {
    pub name: String,
    pub category: Option<String>,
    pub contact: Option<String>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub address: Option<String>,
    pub bank_account: Option<String>,
    pub bank_name: Option<String>,
    pub tax_number: Option<String>,
    pub credit_code: Option<String>,
    pub registered_address: Option<String>,
    pub business_scope: Option<String>,
    pub tax_type: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdatePartner {
    pub id: i64,
    pub name: Option<String>,
    pub category: Option<String>,
    pub contact: Option<String>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub address: Option<String>,
    pub bank_account: Option<String>,
    pub bank_name: Option<String>,
    pub tax_number: Option<String>,
    pub credit_code: Option<String>,
    pub registered_address: Option<String>,
    pub business_scope: Option<String>,
    pub tax_type: Option<String>,
    pub remarks: Option<String>,
}

#[command]
pub fn get_partners(state: State<'_, AppState>, project_id: Option<i64>) -> AppResult<Vec<Partner>> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let sql = if project_id.is_some() {
        "SELECT id, name, category, contact, phone, email, address, bank_account, bank_name,
                tax_number, credit_code, registered_address, business_scope, tax_type,
                license_file, license_file_type, other_files, other_files_type,
                project_ids, remarks, created_at, updated_at
         FROM partners WHERE project_ids LIKE ?1 ORDER BY name"
    } else {
        "SELECT id, name, category, contact, phone, email, address, bank_account, bank_name,
                tax_number, credit_code, registered_address, business_scope, tax_type,
                license_file, license_file_type, other_files, other_files_type,
                project_ids, remarks, created_at, updated_at
         FROM partners ORDER BY name"
    };

    let mut stmt = db.prepare(sql).map_err(|e| AppError::Database(e.to_string()))?;

    let partners = if let Some(pid) = project_id {
        let pattern = format!("%{}%", pid);
        stmt.query_map(params![pattern], |row| {
            Ok(Partner {
                id: row.get(0)?,
                name: row.get(1)?,
                category: row.get(2)?,
                contact: row.get(3)?,
                phone: row.get(4)?,
                email: row.get(5)?,
                address: row.get(6)?,
                bank_account: row.get(7)?,
                bank_name: row.get(8)?,
                tax_number: row.get(9)?,
                credit_code: row.get(10)?,
                registered_address: row.get(11)?,
                business_scope: row.get(12)?,
                tax_type: row.get(13)?,
                license_file: row.get(14)?,
                license_file_type: row.get(15)?,
                other_files: row.get(16)?,
                other_files_type: row.get(17)?,
                project_ids: parse_ids(row.get(18)?),
                remarks: row.get(19)?,
                created_at: row.get(20)?,
                updated_at: row.get(21)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?
    } else {
        stmt.query_map([], |row| {
            Ok(Partner {
                id: row.get(0)?,
                name: row.get(1)?,
                category: row.get(2)?,
                contact: row.get(3)?,
                phone: row.get(4)?,
                email: row.get(5)?,
                address: row.get(6)?,
                bank_account: row.get(7)?,
                bank_name: row.get(8)?,
                tax_number: row.get(9)?,
                credit_code: row.get(10)?,
                registered_address: row.get(11)?,
                business_scope: row.get(12)?,
                tax_type: row.get(13)?,
                license_file: row.get(14)?,
                license_file_type: row.get(15)?,
                other_files: row.get(16)?,
                other_files_type: row.get(17)?,
                project_ids: parse_ids(row.get(18)?),
                remarks: row.get(19)?,
                created_at: row.get(20)?,
                updated_at: row.get(21)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?
    };

    Ok(partners)
}

#[command]
pub fn create_partner(state: State<'_, AppState>, partner: CreatePartner) -> AppResult<i64> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;
    let now = chrono::Local::now().format("%Y-%m-%d %H:%M:%S").to_string();

    db.execute(
        "INSERT INTO partners (name, category, contact, phone, email, address,
         bank_account, bank_name, tax_number, credit_code, registered_address,
         business_scope, tax_type, created_at, updated_at)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14, ?15)",
        params![
            partner.name,
            partner.category,
            partner.contact,
            partner.phone,
            partner.email,
            partner.address,
            partner.bank_account,
            partner.bank_name,
            partner.tax_number,
            partner.credit_code,
            partner.registered_address,
            partner.business_scope,
            partner.tax_type,
            now,
            now,
        ],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    Ok(db.last_insert_rowid())
}

#[command]
pub fn update_partner(state: State<'_, AppState>, partner: UpdatePartner) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;
    let now = chrono::Local::now().format("%Y-%m-%d %H:%M:%S").to_string();

    let affected = db
        .execute(
            "UPDATE partners SET name = COALESCE(?1, name), category = COALESCE(?2, category),
             contact = COALESCE(?3, contact), phone = COALESCE(?4, phone),
             email = COALESCE(?5, email), address = COALESCE(?6, address),
             bank_account = COALESCE(?7, bank_account), bank_name = COALESCE(?8, bank_name),
             tax_number = COALESCE(?9, tax_number), credit_code = COALESCE(?10, credit_code),
             registered_address = COALESCE(?11, registered_address),
             business_scope = COALESCE(?12, business_scope), tax_type = COALESCE(?13, tax_type),
             remarks = COALESCE(?14, remarks), updated_at = ?15
             WHERE id = ?16",
            params![
                partner.name,
                partner.category,
                partner.contact,
                partner.phone,
                partner.email,
                partner.address,
                partner.bank_account,
                partner.bank_name,
                partner.tax_number,
                partner.credit_code,
                partner.registered_address,
                partner.business_scope,
                partner.tax_type,
                partner.remarks,
                now,
                partner.id,
            ],
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("合作伙伴 {} 不存在", partner.id)));
    }

    Ok(())
}

#[command]
pub fn delete_partner(state: State<'_, AppState>, id: i64) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute("DELETE FROM partners WHERE id = ?1", params![id])
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("合作伙伴 {} 不存在", id)));
    }

    Ok(())
}
